java/calcite/src/main/java/org/apache/calcite/prepare/MapDPlanner.java

```diff
@@ -192,6 +210,42 @@ private RelRoot applyQueryOptimizationRules(RelRoot root) {
     return root.withRel(rootRelNode);
   }
 
+  public RelRoot optimizeRaQuery(String query, MapDSchema schema) throws IOException {
+    ready();
+    RexBuilder builder = new RexBuilder(getTypeFactory());
+    RelOptCluster cluster = RelOptCluster.create(new VolcanoPlanner(), builder);
+    CalciteCatalogReader catalogReader = createCatalogReader();
+    MapDRelJsonReader reader = new MapDRelJsonReader(cluster, catalogReader, schema);
+
+    RelRoot relR = RelRoot.of(reader.read(query), SqlKind.SELECT);
+    applyQueryOptimizationRules(relR);
+    applyFilterPushdown(relR);
+
+    ProjectMergeRule projectMergeRule =
+            new ProjectMergeRule(true, RelFactories.LOGICAL_BUILDER);
+    final Program program =
+            Programs.hep(ImmutableList.of(FilterProjectTransposeRule.INSTANCE,
+                                 projectMergeRule,
+                                 ProjectProjectRemoveRule.INSTANCE,
+                                 FilterMergeRule.INSTANCE),
+                    true,
+                    DefaultRelMetadataProvider.INSTANCE);
+    RelNode oldRel;
+    RelNode newRel = relR.project();
+
+    do {
+      oldRel = newRel;
+      newRel = program.run(null,
+              oldRel,
+              null,
+              ImmutableList.<RelOptMaterialization>of(),
+              ImmutableList.<RelOptLattice>of());
+      // there must be a better way to compare these
```

Comment:
This looks like legacy code -- we have since modified some of these optimizations to be a bit cleaner and use built in Calcite rules, see https://github.com/omnisci/omniscidb/blob/master/java/calcite/src/main/java/com/mapd/calcite/parser/MapDParser.java#L821

Reply:
I see, thanks. I think, I would better send the fix as separate pr, as we need to spend some time testing new logic.